Changing the SegFile class

spxiwh opened the pull request:
***I am still testing this, please don't merge yet***

Hi Alex,

I wanted to put this change out there now. This completes the rework of the SegFile class as we have been discussing and moves the various workflow generators around it. I have verified that the PSD and HDF coinc workflows generate successfully with the new code, but I still need to check GRB (I don't care about the old generator yet, as it will be changed quite a bit when we finish addressing the segment-related changes #127).

At this point I wanted to get feedback from you, or anyone else, if you are unhappy with this, or want more features.

Assuming we don't break anything, these changes should be invisible to the end user, but should make it easier for those of us writing workflows to deal with segment files in workflows.

Cheers
Ian

ahnitz commented:
Ian, I don't see anything obvious I object to when scanning through the diffs. The only things I don't like really are the ones you've already commented that require us to also use Larne's new file class interface, which will come in a separate pull request.

spxiwh commented:
Hi Alex,

I've just removed the on hold label, and would now like to start getting this onto master. I have rebased the segment_work branch to master and ran Duncan's test workflow (updating a few things that were broken), so that it now runs:

https://galahad.aei.mpg.de/~spxiwh/LVC/aLIGO/O1/analyses/workflow_test/

The only thing I'm not really happy with is that in pycbc/events/veto.py the get_segment_definer_comments function used to return IFO:NAME:VERSION as the segment name. However, as the VERSION is not really used anywhere other than the veto definer, the SegFile class doesn't use this, and so IFO:NAME is generally the key used for segments. At the moment I allowed this function to return either format, in case this is being used with veto-definer files. The other option is to add VERSION throughout the SegFile and everywhere it is used, but I think this will be painful to always remember to join "0" to the name and the ifo.

Anyway, at the moment this works for the standard workflow and I would like to merge. Edits to the segment.py module in workflow.py will come next. Both this patch, and the next, change quite a bit about how we use segments. People may find issues in code outside of the standard workflow. If reported I can fix these quickly, or people can just use the O1 release branch. I think we've held off some of these changes long enough already!

Cheers
Ian

cmbiwer commented:
>> The only thing I'm not really happy with is that in pycbc/events/veto.py the get_segment_definer_comments function used to return IFO:NAME:VERSION as the segment name. However, as the VERSION is not really used anywhere other than the veto definer, the SegFile class doesn't use this, and so IFO:NAME is generally the key used for segments. At the moment I allowed this function to return either format, in case this is being used with veto-definer files. The other option is to add VERSION throughout the SegFile and everywhere it is used, but I think this will be painful to always remember to join "0" to the name and the ifo.

The executable ``pycbc_page_vetotable`` that prints the veto definer file is using this. And version is important in that case (at least for O1) because we have the same flag in the veto-definer file multiple times with different versions. Whether that's best practice is another question but its there.

spxiwh commented:
Hi Chris,

To clarify pycbc_page_vetotable works as before:

https://galahad.aei.mpg.de/~spxiwh/LVC/aLIGO/O1/analyses/workflow_test/1._analysis_time/1.04_veto_definer/

as do the other codes reading segment files.

I think your point though is that this exe doesn't actually use get_segment_definer_comments at all, it just imports it. Do you know of any code that is using the version functionality within get_segment_definer_comments?

Cheers
Ian

cmbiwer commented:
Hey Ian,

I quickly realized it was just being imported. My mistake.

Though I would not expect it to negatively affect the two codes that actually use it (``pycbc_page_segtable`` and ``pycbc_page_segplot``). They use it to print the comments in the caption on the html page. But all those segments are currently versionless anyways.

So in short, removing the version should not have any negative affects.

ahnitz commented:
It will take me a little while to digest everything, but first thing I notice is that the extension option seems to be inconsistent with the File class. In particular I see instances that use ".xml" and "xml".

spxiwh commented:
I think the extension option goes directly to the File class, and I think that supports ".xml" or "xml" as the same thing ... but we can try to unify that.

ahnitz commented on bin/hdfcoinc/pycbc_make_psd_estimation_workflow:
```diff
@@ -103,9 +101,11 @@ for ifo, segments in science_segs.items():
-    broken_segments_file = segments_to_file(
-            broken_segments, broken_segments_path, 'SCIENCE_OK_BROKEN', ifo=ifo)
+    broken_segments_file = pycbc.workflow.SegFile.from_segment_list(\
+                     'SCIENCE_OK_BROKEN', broken_segments, 'SCIENCE_OK_BROKEN',
```
I don't think you need a \ after a parenthesis.

ahnitz commented:
Ok, if they are the same thing, it is OK.

spxiwh commented:
```
    def _filename(self, ifo, description, extension, segment):
        """
        Construct the standard output filename. Should only be used internally
        of the File class.
        """
        if extension.startswith('.'):
            extension = extension[1:]
```

ahnitz commented:
Ok, good. So either works fine then.

On Wed, Feb 10, 2016 at 2:13 PM, spxiwh <notifications@github.com> wrote:

>     def _filename(self, ifo, description, extension, segment):
>         """
>         Construct the standard output filename. Should only be used internally
>         of the File class.
>         """
>         if extension.startswith('.'):
>             extension = extension[1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ligo-cbc/pycbc/pull/408#issuecomment-182368102>.
>



-- 
Dr. Alexander Nitz
Max Planck Institute for Gravitational Physics (Albert Einstein Institute)
Callinstrasse 38
D-30167 Hannover,  Germany
Tel: +49 511 762-17097

spxiwh commented:
Hi Alex, I've removed the backslash and remind that I have run the test workflow end-to-end with this, so the ".xml" vs "xml" has been shown to work fine .... That inconsistency was actually there *before* this patch.

ahnitz commented:
Ok, that sounds fine to me then.

On Wed, Feb 10, 2016 at 2:59 PM, spxiwh <notifications@github.com> wrote:

> Hi Alex, I've removed the backslash and remind that I have run the test
> workflow end-to-end with this, so the ".xml" vs "xml" has been shown to
> work fine .... That inconsistency was actually there *before* this patch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ligo-cbc/pycbc/pull/408#issuecomment-182387204>.
>



-- 
Dr. Alexander Nitz
Max Planck Institute for Gravitational Physics (Albert Einstein Institute)
Callinstrasse 38
D-30167 Hannover,  Germany
Tel: +49 511 762-17097

ahnitz commented:
I'll merge once the standard travis checks complete in ~ 20 minutes

spxiwh commented:
Okay, I'll merge this once the sanity tests complete then unless there are last-minute objections.
